### ISSUE TITLE
[FSDP] Add `foreach` support to `FSDP.clip_grad_norm_()`

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#92736]
See https://github.com/pytorch/pytorch/pull/91846 for `torch.nn.utils.clip_grad_norm_()` changes. Since `FSDP.clip_grad_norm_()` has to re-implement the logic to include gradient unsharding, we have to manually upstream improvements like this.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501

**Comment by awgu** [comment]
`foreach` support is not so important for `use_orig_params=False`. For `True`, it is still not so important especially at large scale since each rank likely only has a small subset of parameters with > 0 numel.
